scala-api/src/main/scala/com/cloudera/livy/scalaapi/ScalaJobHandle.scala

```diff
@@ -0,0 +1,123 @@
+
+import scala.concurrent.{CanAwait, ExecutionContext, Future, TimeoutException}
+import scala.concurrent.duration.Duration
+import scala.util.Try
+
+import com.cloudera.livy.JobHandle
+import com.cloudera.livy.JobHandle.{Listener, State}
+
+class ScalaJobHandle[T] private[livy] (jobHandle: JobHandle[T]) extends Future[T] {
+
+  def state: State = jobHandle.getState()
+
+  override def onComplete[U](func: (Try[T]) => U)(implicit executor: ExecutionContext): Unit = {
+    jobHandle.addListener(new AbstractScalaJobHandleListener[T] {
+      override def onJobSucceeded(job: JobHandle[T], result: T): Unit = {
+        val onJobSucceededTask = new Runnable {
+          override def run(): Unit = func(Try(result))
+        }
+        executor.execute(onJobSucceededTask)
+      }
+
+      override def onJobFailed(job: JobHandle[T], cause: Throwable): Unit = {
+        val onJobFailedTask = new Runnable {
+          override def run(): Unit = func(Try(getJavaFutureResult(job)))
+        }
+        executor.execute(onJobFailedTask)
+      }
+    })
+  }
+
+  def onJobQueued[U](func: => Unit)(implicit executor: ExecutionContext): Unit = {
+    jobHandle.addListener(new AbstractScalaJobHandleListener[T] {
+      override def onJobQueued(job: JobHandle[T]): Unit = {
+        val onJobQueuedTask = new Runnable {
+          override def run(): Unit = func
+        }
+        executor.execute(onJobQueuedTask)
+      }
+    })
+  }
+
+  def onJobStarted[U](func: => Unit)(implicit executor: ExecutionContext): Unit = {
+    jobHandle.addListener(new AbstractScalaJobHandleListener[T] {
+      override def onJobStarted(job: JobHandle[T]): Unit = {
+        val onJobStartedTask = new Runnable {
+          override def run(): Unit = func
+        }
+        executor.execute(onJobStartedTask)
+      }
+    })
+  }
+
+  def onJobCancelled[U](func: Boolean => Unit)(implicit executor: ExecutionContext): Unit = {
+    jobHandle.addListener(new AbstractScalaJobHandleListener[T] {
+      override def onJobCancelled(job: JobHandle[T]): Unit = {
+        val onJobCancelledTask = new Runnable {
+          override def run(): Unit = func(job.cancel(false))
+        }
+        executor.execute(onJobCancelledTask)
+      }
+    })
+  }
+
+  override def isCompleted: Boolean = jobHandle.isDone
+
+  override def value: Option[Try[T]] = {
+    if (isCompleted) {
+      Some(Try(getJavaFutureResult(jobHandle)))
+    } else {
+      None
+    }
+  }
+
+  @throws(classOf[Exception])
+  override def result(atMost: Duration)(implicit permit: CanAwait): T =
+    getJavaFutureResultAfterGivenWaitTime(atMost)
+
+  @throws(classOf[InterruptedException])
+  @throws(classOf[TimeoutException])
+  override def ready(atMost: Duration)(implicit permit: CanAwait): ScalaJobHandle.this.type = {
+    getJavaFutureResultAfterGivenWaitTime(atMost)
+    this
+  }
+
+  private def getJavaFutureResultAfterGivenWaitTime(atMost: Duration): T = {
+    if (!atMost.isFinite()) {
```

Comment:
`getJavaFutureResult` already handles this right? No need to duplicate the check. Then you don't need this method.